src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      newState = Object.assign(newState, action.extraData);
+    }
+
+    if (action.type === 'removeProperties') {
+      const keysToRemove = action.keysToRemove || [];
+
+      for (const key of keysToRemove) {
+        delete newState[key];
+      }
+    }
+
+    if (action.type === 'clear') {
+      newState = {};
+    }
+
+    result.push(newState);
+    currentState = newState;
```

Comment:
The reassigning of currentState to newState is not needed as you are already updating currentState inside each action type.